### PR TITLE
vim-patch:9.0.{1585,1595}

### PR DESCRIPTION
**Pull request by luukvbaal**
**vim-patch:9.0.1585: weird use of static variables for spell checking**

Problem:    Weird use of static variables for spell checking.
Solution:   Move the variables to a structure and pass them from win_update()
            to win_line(). (Luuk van Baal, closes vim/vim#12448)

https://github.com/vim/vim/commit/30805a1aba0067cf0087f9a0e5c184562433e2e7

**vim-patch:9.0.1595: line pointer becomes invalid when using spell checking**

Problem:    Line pointer becomes invalid when using spell checking.
Solution:   Call ml_get() at the right places. (Luuk van Baal, closes vim/vim#12456)

https://github.com/vim/vim/commit/e84c773d42e8b6ef0f8ae9b6c7312e0fd47909af

### CODE REVIEW

File: src/nvim/drawline.c [hunk 1]
```diff
@@ -1021,21 +1021,25 @@ static void win_line_continue(winlinevars_T *wlv)
   }
 }
 
-/// Display line "lnum" of window 'wp' on the screen.
+/// Display line "lnum" of window "wp" on the screen.
+/// Start at row "startrow", stop when "endrow" is reached.
+/// When "number_only" is TRUE only update the number column.
```
Comment (zeertzjq) [inline src/nvim/drawline.c:1026]:
```suggestion
/// When "number_only" is true only update the number column.
```

**Review by zeertzjq — CHANGES_REQUESTED** [review]
You need to pass a zero `spellvars_T` on folded lines and end filler lines. I'm surprised that there are no compiler or clang-tidy warnings for passing `false` as a pointer.

**Comment by luukvbaal** [comment]
Quite surprising indeed.

**Comment by zeertzjq** [comment]
When I run `test/functional/ui/fold_spec.lua` tests locally on this PR the test `folded lines without ext_multigrid work with spell` fails, but it passes on CI. Strange.

**Comment by luukvbaal** [comment]
Right, guess I'll make another patch to vim first to fix this.

**Comment by zeertzjq** [comment]
I think it works fine in Vim.

**Comment by luukvbaal** [comment]
`line` is cleared by an `ml_get_buf()` call that this patch added.
I think it should be restored by calling `ml_get_buf()` again.

**Comment by zeertzjq** [comment]
Oh, `test_override('alloc_lines', 1)` isn't called in Vim instances run with `RunVimInTerminal()`.

File: src/nvim/drawline.c [hunk 2]
```diff
@@ -1021,21 +1021,25 @@ static void win_line_continue(winlinevars_T *wlv)
   }
 }
 
-/// Display line "lnum" of window 'wp' on the screen.
+/// Display line "lnum" of window "wp" on the screen.
+/// Start at row "startrow", stop when "endrow" is reached.
+/// When "number_only" is true only update the number column.
+/// "spv" is used to store information for spell checking, kept between
+/// sequential calls for the same window.
```
Comment (luukvbaal) [inline src/nvim/drawline.c:1028]: Is this desired? I noticed it was missing but I guess it is included in short-form in the @param descriptions.

Comment (zeertzjq) [reply src/nvim/drawline.c:1028]: What was missing?

Comment (luukvbaal) [reply src/nvim/drawline.c:1028]: (Everything added here?) I mean I noticed that the function comment for our `win_line()` was much shorter. But I guess the information is included in the param comments instead.
Do we want both or should I remove it again?

Comment (zeertzjq) [reply src/nvim/drawline.c:1028]: I think they should be moved to the param comments.

File: src/nvim/drawline.c [hunk 3]
```diff
@@ -1420,18 +1377,46 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int mod_top, bo
     line_attr_lowprio_save = wlv.line_attr_lowprio;
   }
 
-  line = end_fill ? "" : ml_get_buf(wp->w_buffer, lnum, false);
-  ptr = line;
+  if (spv->spv_has_spell && !number_only) {
```
Comment (luukvbaal) [inline src/nvim/drawline.c:1380]: `number_only`, `end_fill` and `has_fold`, are used quite inconsistently throughout `win_line()`. I think a lot of times if one is used to guard a block, the others should be as well. That is not always the case currently.

Might verify that this doesn't break anything and/or introduce a new boolean to skip these blocks later...

Comment (zeertzjq) [reply src/nvim/drawline.c:1380]: In other places this seems already guarded by other things like `v > 0`.

**Review by zeertzjq — APPROVED** [review]

File: src/nvim/drawline.c (hunk 3 above)
Comment (luukvbaal) [reply src/nvim/drawline.c:1380]: Why not just add `&& !has_fold && !end_fill` to the spell check block?

Comment (zeertzjq) [reply src/nvim/drawline.c:1380]: It matches the old code better and make it less necessary to decide whether to add `&& !has_fold && !end_fill` in other places where `spv_has_spell` appears.

**Pull request merged**